src/libraries/System.Runtime/ref/System.Runtime.cs

```diff
@@ -10246,6 +10246,8 @@ public sealed partial class BufferedStream : System.IO.Stream
     }
     public static partial class Directory
     {
+        public static bool Copy(string sourcePath, string destinationPath, bool recursive) { throw null; }
```

Comment:
@Jozkee @adamsitnik @bartonjs @danmoseley @stephentoub this API fails in two different ways: by returning `false`, or by throwing `IOException`.

Can the user act upon this returing `false` in some meaningful way?
The `false` hides more information which would be conveyed in an `IOException`, like what path failed to read and why.
Splitting the error cases also makes the implementation more complex.

Can we change the signature to `void` and always throw `IOException`?

Reply:
@Symbai it is an attempt to provide a best-effort copy where the user has no idea what has been copied and why the copy failed. That means it is hard to handle the `false` case.

I think it would be better to let the user roll their own best-effort copy and nail the semantics as they want them.

Most users for this API won't be looking for a best-effort copy, so this API forces them to deal with a `false` path they don't want.

Reply:
Well it has been discussed in API review and this is how it got approved and so it should get implemented. As for the pro and cons and why (arguments) please watch the API review video. There were also example situations discussed where returning `false` makes more sense than throwing. Also please don't speak for "Most users", as I for once don't agree.

Reply:
@Symbai I watched the API review. I think the `bool`-return got added under the false presumption it was needed to know everything was copied.
Also in the review it gets mentioned there shouldn't be a ton of options, and if you want more specific behavior, you can do it enumerate yourself.
This `bool` return value represents the option 'SkipReadErrors' which every user of this API is now forced to handle as a specific case.

Reply:
> I think the `bool`-return got added under the false presumption it was needed to know everything was copied.

Except for "false presumption", like I said yes. Also, correct me if I am wrong since you just watched it, in the API review it was said that it should enumerate directory by directory and copy file by file and skip only those where it fails but it should NOT throw. The last time I've checked it this implementation doesn't fully meet this criteria, instead of enumerates all directories at once and it throws an exception on fail.

Reply:
@terrajobst ptal.

Reply:
@terrajobst friendly ping

Reply:
What if it fails to copy because the previously enumerated file no longer exists but everything else can be copied? It is ridiculous to say abort everything on the first exception. You won't solve any "incomplete state" because you already have the mess that some files are copied but others not. This IS an incomplete state and you have no idea what the method has copied so far. So there is nothing a developer can react to in your case as well.  Especially when you overwrite files. And most people want to copy everything which can be copied rather than catch - handle - redo - catch - handle - redo - catch - handle - redo and so on. As a side note I don't quite understand why an already reviewed API is suddenly getting questioned because people forgot was has been discussed and reviewed so far, this seems weird. Isn't the API review there to discuss HOW it should get implemented and that everyone in the API review team have agreed to it?

Reply:
> What if it fails to copy because the previously enumerated file no longer exists but everything else can be copied?

Then you have a race condition that's causing the operation to fail.

> It is ridiculous to say abort everything on the first exception

It is not "ridiculous".

> So there is nothing a developer can react to in your case as well.

The _user_ can, and they have no idea what to react to if there's no error details.

> And most people want to copy everything which can be copied rather than catch - handle - redo - catch - handle - redo - catch - handle - redo and so on.

Based on what research?

> As a side note I don't quite understand why an already reviewed API is suddenly getting questioned because people forgot was has been discussed and reviewed so far, this seems weird.

I did not forget.  I watched the video.  I disagree with how we ended up here.  And an API isn't done until it's shipped.  Just because it was approved one way in no way ties our hands.

Reply:
My recollection (no, I haven't rewatched the video) is that this is supposed to be a friendly method.  Basically, on the order of "back up this directory to there".  If it returns true, everything's happy.  If it returns false then now I have to do something more complicated.  The copy stopping at the moment the first thing went wrong is super easy to explain, sure, but has low utility... if it long-path-name violated out on the first file but literally everything else could have succeeded, then the method has now annoyed me more than helped me.

Put another way: With no retry policy/error handler callback/etc, the method feels, to me, too trivial to be worth adding if it throws on the first failure (except for a few degenerate cases, like "your destination is an illegal path").

Reply:
> If it returns false then now I have to do something more complicated

Do what?  How do you know what went wrong and what to do differently next time / how to fix it so it doesn't happen again?

> With no retry policy/error handler callback/etc, the method feels, to me, too trivial to be worth adding if it throws on the first failure (except for a few degenerate cases, like "your destination is an illegal path").

I agree with that.  I also think it's not worth adding if it only does some of what you asked it to do and happily returns false with no additional information.

Reply:
Could we have a policy so we can satisfy several behaviors, including return false on errors and throw?
Something like:
```cs
public enum DirectoryCopyErrorPolicy
{
    Warn, // Best effort to copy, but return false if something failed.
    Throw, // throws an aggregate exception.
    ThrowFirst, // throws the first exception it gets.
}
```

I don't remember if we discussed this in API review.

Reply:
Maybe the issue should go back to API review? And we should close this PR for now?

Reply:
Or it should get implemented on how it was reviewed and improvements/changes are tracked in a separate issue. Wouldn't even be the first time.

Reply:
I think this should get back to APi review.